### ISSUE TITLE
How to change negotiation type dynamicly via eureka meta data. 

### ISSUE THREAD

**Issue by EdAndrew** [grpc/grpc-java#9064]
**The context**

I want to make new feature about grpc channel, and it is :

Grpc server can write meta data in eureka about whether negotiation type client should use to communicate with server. And before requests grpc server, grpc client should read this meta data from eureka and change own negotiotion type in channel.

**The question**

I know grpc client can use `NameResolver` to change ips of grpc server dynamicly, and want to know how to rewrite code to impletaion dynamic negotation type. Negotiation type is configured in Spring Boot init stage, how to change it in communication state?

Thanks for you answer!

**Comment by ejona86** [comment]
This looks completely fine within a NameResolver. NameResolvers can look up IPs from a source of their choosing, which means they can also do a level of indirection and look up results in Eureka and then use some secondary resolution type based on the results.

We can't answer from a Spring perspective.

**Comment by ST-DDT** [comment]
Let me rephrase the question.
How can I use a `NameResolverProvider` to enable/disable TLS for some of the addresses.

Information resolved from whatever source:

| IP:Port | TLS |
| --- | --- |
| 10.0.0.1:80 | false |
| 10.0.0.2:80 | false |
| 10.0.0.3:443 | true |

Is there some attribute, I can set?

````java
new EquivalentAddressGroup(new InetSocketAddress(host, port), Attributes.newBuilder().set("TLS", true).build())
````

**Comment by ejona86** [comment]
Oh, _that_ negotiation type. You have to get fancy and use internal APIs. This is [something that GoogleDefaultChannelCredentials do](https://github.com/grpc/grpc-java/blob/584622c5fa125b327e90f8ecb0c01de5544d39c8/alts/src/main/java/io/grpc/alts/internal/AltsProtocolNegotiator.java#L296). But it requires making your own protocol negotiator which is very unstable and we'd really want such things to be in our repo so we can upgrade usages at the same time we break APIs.

**Comment by ejona86** [comment]
Seems like this is "resolved," as in, "no, you can't really do want you want." It is possible, but not practical. If that doesn't quite sound right or I'm missing another part of this question (or you want to learn more about the shenanigans to get this to happen), comment, and it can be reopened.
